### DISCUSSION TITLE
Problem with run_javascript and 'getElement'

### DISCUSSION THREAD

**Discussion by ed2050** [zauberzeug/nicegui#2058] [Q&A] [answered]
### Question

I'm having trouble with using getElement in run_javascript.  Not sure where the problem is.

I'm trying to scroll an element on the page into view.  My code makes this call:

```py
await ui.run_javascript (f'''
    console.log ('scroll to id = {node.id}') ;
    const element = getElement ({node.id}).$refs.qRef.getNativeElement () ;
    element.scrollIntoView () ;
''', respond = false)
```

It's the same js code proposed by @falkoschindler [here](https://github.com/zauberzeug/nicegui/discussions/1218#discussioncomment-6511278), but calling scrollIntoView instead of focus.

Above call produces this error in browser console (no messages in nicegui terminal) : 
```sh
scroll to id = 18
VM20755:3 Uncaught (in promise) TypeError: Cannot read properties of undefined (reading 'qRef')
    at eval (eval at <anonymous> ((index):136:42), <anonymous>:3:41)
    at (index):347:42
    at new Promise (<anonymous>)
    at runJavascript ((index):347:10)
    at run_javascript ((index):450:38)
    at (index):463:45
    at a.<anonymous> ((index):469:27)
    at j.emit (index.mjs:136:20)
    at a.value (socket.js:498:20)
    at a.value (socket.js:485:18)
```

I got the focus () version working elsewhere, see #2044.  It seems the qRef object doesn't exist here somehow?

## Context
I'm implementing keyboard shortcuts to step through items.  RightArrow for next, LefArrow for previous.  The stepping works fine.  But sometimes next item is off screen.  Hence scrollIntoView.

Code
```py
# g.walkthru is a list of image elements
# g.walkpos is position (index) of current element in walkthru

async \
def show_item (pos) :
    node = g.walkthru [pos]
    node.visible = true
    await ui.run_javascript (f'''
        console.log ('scroll to id = {node.id}') ;
        const element = getElement ({node.id}).$refs.qRef.getNativeElement () ;
        element.scrollIntoView () ;
    ''', respond = false)

# ------------------------------------------------

async \
def show_prev () :
    g.walkpos = (g.walkpos - 1) % (len (g.walkthru) or 1)
    show_item (g.walkpos)

# ------------------------------------------------

async \
def show_next () :
    g.walkpos = (g.walkpos + 1) % (len (g.walkthru) or 1)
    show_item (g.walkpos)

# ------------------------------------------------

async \
def key_listen (e) :

    DEBUG and ui.notify (f'got key : { e.key }')
    if not e.action.keydown :  return

    if e.key == 'ArrowRight' :
        await show_next ()

    elif e.key == 'ArrowLeft' :
        await show_prev ()

# ------------------------------------------------

ui.keyboard (on_key = key_listen)
```

## Possible Causes

1. Could it be a client.connected issue?  Working example with focus called `client.connected ()` first.  Problem is, this code is invoked from a keyboard listener.  I don't see how to get a client object passed to then call connected () on.  It doesn't seem to be the problem since I get an error in browser console.
2. Could it be an async issue?  I originally had it without async / await everywhere.  Then nicegui throws an error saying that run_javascript was never awaited.  1.4 release changes this so run_javascripts doesn't have to be awaited, but I'm on 1.3.16.  Again I don't think this is the problem since it gets to actually running js.

Any other ideas?  Should I resort to assigning my own id to each walkthru image, then calling `document.getElementById (imageid).scrollIntoView`?

Thanks for any suggestions.

**Comment by ed2050** [comment]
Update: I got it to work by adding my own id with `ui.image.props (f'id={ myid }')` and then calling `document.getElementById (myid).scrollIntoView ()`.

Still want to know why the qref approach failed.  I'm sure it will come up again.

**Comment by natankeddem** [comment]
Related?
https://github.com/zauberzeug/nicegui/issues/1507

**Reply by ed2050** [reply]
Interesting.  Good to know that getElement works for some nicegui element types and not others.  I'll avoid it in future.  Thanks!

**Comment by falkoschindler** [comment, marked as answer]
Here is a possible solution:

```py
images = [ui.image(f'https://picsum.photos/id/{i}/640/360').classes('w-80') for i in range(10)]

async def show_image(index: int) -> None:
    await ui.run_javascript(f'getElement({images[index].id}).$el.scrollIntoView()')

with ui.left_drawer():
    ui.button('show first image', on_click=lambda: show_image(0))
    ui.button('show last image', on_click=lambda: show_image(-1))
```

**Reply by ed2050** [reply]
Cool, thanks falko.  Perhaps in future I should avoid js call chains with unfamiliar methods. 😄  Vanilla js isn't sexy but it works.  At least until I add brython to my app. 😉
